src/focus-ring.js

```diff
@@ -1,67 +1,112 @@
+        if (proto.matches)
+            return proto.matches;
+        if (proto.matchesSelector)
+            return proto.matchesSelector;
+        if (proto.webkitMatchesSelector)
+            return proto.webkitMatchesSelector;
+        if (proto.mozMatchesSelector)
+            return proto.mozMatchesSelector;
+        if (proto.msMatchesSelector)
+            return proto.msMatchesSelector;
+        console.error('Couldn\'t find a matches method on Element.prototype.');
+    }());
+
+    /**
+     * Computes whether the given element should automatically trigger the
+     * `focus-ring` class being added, i.e. whether it should always match
+     * `:focus-ring` when focused.
+     * @param {Element} el
+     * @return {boolean}
+     */
+    function focusTriggersKeyboardModality(el) {
+        var triggers = false;
+        if (matcher) {
+            triggers = matcher.call(el, keyboardModalityWhitelist) &&
+                       matcher.call(el, ':not([readonly]');
+        }
+        return triggers;
+    }
+
+    /**
+     * Add the `focus-ring` class to the given element if it was not added by
+     * the author.
+     * @param {Element} el
+     */
+    function addFocusRingClass(el) {
+        if (el.classList.contains('focus-ring'))
+            return;
+        el.classList.add('focus-ring');
+        el.setAttribute('data-focus-ring-added', '');
+    }
+
+    /**
+     * Remove the `focus-ring` class from the given element if it was not
+     * originally added by the author.
+     * @param {Element} el
+     */
+    function removeFocusRingClass(el) {
+        if (!el.hasAttribute('data-focus-ring-added'))
+            return;
+        el.classList.remove('focus-ring');
+        el.removeAttribute('data-focus-ring-added')
+    }
 
-    document.body.addEventListener('keydown', function() {
+    /**
+     * On `keydown`, set `hadKeyboardEvent`, to be removed 100ms later if there
+     * are no further keyboard events.  The 100ms throttle handles cases where
+     * focus is redirected programmatically after a keyboard event, such as
+     * opening a menu or dialog.
+     */
+    function onKeyDown() {
         hadKeyboardEvent = true;
+        // `activeElement` defaults to document.body if nothing focused,
+        // so check the active element is actually focused.
         if (document.activeElement.matches(':focus')) {
             addFocusRingClass(document.activeElement);
         }
-        if (isHandlingKeyboardThrottle) {
+        if (keyboardThrottleTimeoutID !== 0) {
             clearTimeout(isHandlingKeyboardThrottle);
         }
-        isHandlingKeyboardThrottle = setTimeout(function() {
+        keyboardThrottleTimeoutID = setTimeout(function() {
             hadKeyboardEvent = false;
+            keyboardThrottleTimeoutID = 0;
         }, 100);
-    }, true);
+    }
 
-    document.body.addEventListener('focus', function(e) {
+    /**
+     * On `focus`, add the `focus-ring` class to the target if necessary.
```

Comment:
replace `if necessary` with `if the element matches a selector in the keyboardModalityWhitelist whitelist`